compiler/passes/scopeResolve.cpp

```diff
@@ -1961,140 +1976,285 @@ bool lookupThisScopeAndUses(const char*           name,
+
   if (symbols.size() == 0) {
-    // Nothing found so far, look into the uses.
-    if (BlockStmt* block = toBlockStmt(scope)) {
-      if (block->useList != NULL) {
-        Vec<VisibilityStmt*>* moduleUses = NULL;
-
-        if (moduleUsesCache.count(block) == 0) {
-          moduleUses = new Vec<VisibilityStmt*>();
-
-          for_actuals(expr, block->useList) {
-            // Ensure we only have use or import statements in this list
-            if (UseStmt* use = toUseStmt(expr)) {
-              moduleUses->add(use);
-            } else if (ImportStmt* import = toImportStmt(expr)) {
-              moduleUses->add(import);
-            } else {
-              INT_FATAL("Bad contents for useList, expected use or import");
-            }
-          }
+    // check private use only, not including names of modules
+    // (this forms the 1st shadow scope for a private use)
+    visitedModules.clear();
+    lookupUseImport(name, context, scope, symbols,
+                    renameLocs, storeRenames, reexportPts, visitedModules,
+                    /* forShadowScope */ true,
+                    /* publicOnly */ false);
+  }
 
-          INT_ASSERT(moduleUses->n);
+  if (symbols.size() == 0) {
+    // Check to see if the module name matches what is use'd
+    // with a private use
+    // (this forms the 2nd shadow scope for a private use)
+    lookupUsedImportedMod(name, context, scope, symbols,
+                          renameLocs, storeRenames,
+                          /* forShadowScope */ true,
+                          /* publicOnly */ false);
+  }
 
-          buildBreadthFirstModuleList(moduleUses);
+  return symbols.size() != 0;
+}
 
-          if (enableModuleUsesCache)
-            moduleUsesCache[block] = moduleUses;
+static void lookupUseImport(const char*           name,
+                            BaseAST*              context,
+                            BaseAST*              scope,
+                            std::vector<Symbol*>& symbols,
+                            std::map<Symbol*, astlocT*>& renameLocs,
+                            bool storeRenames,
+                            std::map<Symbol*, VisibilityStmt*>& reexportPts,
+                            VisitedModulesSet& visitedModules,
+                            bool forShadowScope,
+                            bool publicOnly) {
+  // Nothing found so far, look into the uses.
+  if (BlockStmt* block = toBlockStmt(scope)) {
+    if (block->useList != NULL) {
+
+      // check to see if we have already visited this scope
+      // if it is a module scope (to avoid infinite loop with recursive use)
+      if (ModuleSymbol* mod = toModuleSymbol(block->parentSymbol)) {
+        if (mod->block == block) {
+          // if the module is private, return,
+          // since we can't access anything in it
+          // TODO: does this need to have a more involved check?
+          // could a private outer module use a private inner
+          // module and see it?
```

Comment:
I haven't found a case where this check is wrong. Here was my attempt:

``` chapel
module A {
  private module AImpl {
    var x: int;
  }
  module ASub {
    public use super.AImpl;
  }
}

module Main {
  use A;
  proc main() {
    writeln(x);
  }
}
```

But that works the same way on main and on this branch -- `x` undeclared.

Reply:
Wouldn't that code need a use of ASub to exercise it?

Reply:
Good point. Here is a fixed version:

``` chapel
module A {
  private module AImpl {
    var x: int;
  }
  module ASub {
    public use super.AImpl;
  }
  public use ASub;
}

module Main {
  use A;
  proc main() {
    writeln(x);
  }
}
```

On `main`, this does not compile, with an error that `x` is undeclared. With this PR, it does compile. I think I'll open up an issue about this since it's not completely obvious to me what it should be doing & if we need to fix it, I'd like to let that be follow-up work to this PR. Does that sound OK?

Reply:
I could have sworn we'd already had an issue along those lines

Reply:
I'd appreciate help in finding it, if you are up for that. Anyway is it OK to not fix it for this PR?

Reply:
Yup!  #18249

Reply:
If I'm reading write, the conclusion from that issue is that the new behavior is what we want and this would close #18249.

Reply:
Hmm. It doesn't resolve the original issue there, so what I will do is put a comment there noting that the example above changed behavior.

Reply:
Odd, but that seems like a reasonable course of action to me (and you'll add this test code to the PR, right?)